### PR TITLE
Knocked Preview implementation

### LINKED ISSUE #3420: [Task] Request to join a (public) room
Please see and discuss the details in the meta issue.

### LINKED ISSUE #3421: [Task] Cancel a request to join
Please see and discuss the details in the meta issue.

### PR DESCRIPTION
**Pull request by Velin92**
There is still a bug that prevents knocked room to be fetched properly, that will be fixed later here:
https://github.com/matrix-org/matrix-rust-sdk/pull/4152

fixes #3421
fixes #3420

Also this updates the SDK

https://github.com/user-attachments/assets/e8e93479-f003-400a-89d9-7d37044c492f

### CODE REVIEW

File: ElementX/Sources/Services/Timeline/TimelineProxy.swift [hunk 1]
```diff
@@ -274,7 +274,7 @@ final class TimelineProxy: TimelineProxyProtocol {
                    requestHandle: @MainActor (SendAttachmentJoinHandleProtocol) -> Void) async -> Result<Void, TimelineProxyError> {
         MXLog.info("Sending image")
         
-        let handle = timeline.sendImage(url: url.path(percentEncoded: false), thumbnailUrl: thumbnailURL.path(percentEncoded: false), imageInfo: imageInfo, caption: nil, formattedCaption: nil, progressWatcher: UploadProgressListener { progress in
+        let handle = timeline.sendImage(url: url.path(percentEncoded: false), thumbnailUrl: thumbnailURL.path(percentEncoded: false), imageInfo: imageInfo, caption: nil, formattedCaption: nil, storeInCache: false, progressWatcher: UploadProgressListener { progress in
```
Comment (Velin92) [inline ElementX/Sources/Services/Timeline/TimelineProxy.swift:277]: this is related to the latest SDK update, for now I have set it to false, there are other instances where I did the same, but I am unsure if this is the expected value

File: ElementX/Sources/Services/Timeline/TimelineProxy.swift [hunk 2]
```diff
@@ -316,7 +316,7 @@ final class TimelineProxy: TimelineProxyProtocol {
                    requestHandle: @MainActor (SendAttachmentJoinHandleProtocol) -> Void) async -> Result<Void, TimelineProxyError> {
         MXLog.info("Sending video")
         
-        let handle = timeline.sendVideo(url: url.path(percentEncoded: false), thumbnailUrl: thumbnailURL.path(percentEncoded: false), videoInfo: videoInfo, caption: nil, formattedCaption: nil, progressWatcher: UploadProgressListener { progress in
+        let handle = timeline.sendVideo(url: url.path(percentEncoded: false), thumbnailUrl: thumbnailURL.path(percentEncoded: false), videoInfo: videoInfo, caption: nil, formattedCaption: nil, storeInCache: false, progressWatcher: UploadProgressListener { progress in
```
Comment (Velin92) [inline ElementX/Sources/Services/Timeline/TimelineProxy.swift:319]: as the comment above

File: ElementX/Sources/Services/Timeline/TimelineProxy.swift [hunk 3]
```diff
@@ -340,7 +340,7 @@ final class TimelineProxy: TimelineProxyProtocol {
                           requestHandle: @MainActor (SendAttachmentJoinHandleProtocol) -> Void) async -> Result<Void, TimelineProxyError> {
         MXLog.info("Sending voice message")
         
-        let handle = timeline.sendVoiceMessage(url: url.path(percentEncoded: false), audioInfo: audioInfo, waveform: waveform, caption: nil, formattedCaption: nil, progressWatcher: UploadProgressListener { progress in
+        let handle = timeline.sendVoiceMessage(url: url.path(percentEncoded: false), audioInfo: audioInfo, waveform: waveform, caption: nil, formattedCaption: nil, storeInCache: false, progressWatcher: UploadProgressListener { progress in
```
Comment (Velin92) [inline ElementX/Sources/Services/Timeline/TimelineProxy.swift:343]: as the comment above

File: ElementX/Sources/Services/Timeline/TimelineProxy.swift [hunk 4]
```diff
@@ -227,7 +227,7 @@ final class TimelineProxy: TimelineProxyProtocol {
                    requestHandle: @MainActor (SendAttachmentJoinHandleProtocol) -> Void) async -> Result<Void, TimelineProxyError> {
         MXLog.info("Sending audio")
         
-        let handle = timeline.sendAudio(url: url.path(percentEncoded: false), audioInfo: audioInfo, caption: nil, formattedCaption: nil, progressWatcher: UploadProgressListener { progress in
+        let handle = timeline.sendAudio(url: url.path(percentEncoded: false), audioInfo: audioInfo, caption: nil, formattedCaption: nil, storeInCache: false, progressWatcher: UploadProgressListener { progress in
```
Comment (Velin92) [inline ElementX/Sources/Services/Timeline/TimelineProxy.swift:230]: as the comment above

File: ElementX/Sources/Services/Timeline/TimelineProxy.swift [hunk 5]
```diff
@@ -250,7 +250,7 @@ final class TimelineProxy: TimelineProxyProtocol {
                   requestHandle: @MainActor (SendAttachmentJoinHandleProtocol) -> Void) async -> Result<Void, TimelineProxyError> {
         MXLog.info("Sending file")
         
-        let handle = timeline.sendFile(url: url.path(percentEncoded: false), fileInfo: fileInfo, progressWatcher: UploadProgressListener { progress in
+        let handle = timeline.sendFile(url: url.path(percentEncoded: false), fileInfo: fileInfo, storeInCache: false, progressWatcher: UploadProgressListener { progress in
```
Comment (Velin92) [inline ElementX/Sources/Services/Timeline/TimelineProxy.swift:253]: as the comment above

File: ElementX/Sources/Services/Timeline/TimelineProxy.swift (hunk 1 above)
Comment (stefanceriu) [reply ElementX/Sources/Services/Timeline/TimelineProxy.swift:277]: I raised [a separate PR](https://github.com/element-hq/element-x-ios/pull/3437) for adopting these (based on my existing branch). Let's get that merged and then rebase this PR on top of it.

**Review by stefanceriu — CHANGES_REQUESTED** [review]

File: ElementX/Resources/Localizations/en.lproj/Localizable.strings [hunk 6]
```diff
@@ -352,6 +352,7 @@
 "screen_join_room_knock_message_description" = "Message (optional)";
 "screen_join_room_knock_sent_description" = "You will receive an invite to join the room if your request is accepted.";
 "screen_join_room_knock_sent_title" = "Request to join sent";
+"screen_knocked_knock_sent" = "Request to join sent";
```
[the remark this answers is not in the document]
Comment (Velin92) [reply ElementX/Resources/Localizations/en.lproj/Localizable.strings:355]: reusing the same for the invite which id `screen_invited_invite`, but yeah can change it

File: ElementX/Sources/Services/Room/RoomSummary/RoomSummaryProvider.swift [hunk 7]
```diff
@@ -258,9 +255,15 @@ class RoomSummaryProvider: RoomSummaryProviderProtocol {
         
         let notificationMode = roomInfo.cachedUserDefinedNotificationMode.flatMap { RoomNotificationModeProxy.from(roomNotificationMode: $0) }
         
+        let joinRequestType: RoomSummary.JoinRequestType? = switch roomInfo.membership {
+        case .invited: .invite
+        case .knocked: .knocked
+        default: nil
```
[the remark this answers is not in the document]
Comment (Velin92) [reply ElementX/Sources/Services/Room/RoomSummary/RoomSummaryProvider.swift:261]: the thing is that a room may also be left, joined or simply not joined at all, this property jst tells you that there is currently a join request pending, and if is an invite or a knock

File: ElementX/Sources/Screens/JoinRoomScreen/View/JoinRoomScreen.swift [hunk 8]
```diff
@@ -14,7 +14,7 @@ struct JoinRoomScreen: View {
     @ObservedObject var context: JoinRoomScreenViewModel.Context
     
     var body: some View {
-        FullscreenDialog(topPadding: 80, background: .bloom) {
+        FullscreenDialog(topPadding: context.viewState.mode == .knocked ? 151 : 35, background: .bloom) {
```
[the remark this answers is not in the document]
Comment (Velin92) [reply ElementX/Sources/Screens/JoinRoomScreen/View/JoinRoomScreen.swift:17]: just following the figma redesign values 🤷‍♂️

File: ElementX/Sources/Services/Room/RoomSummary/RoomSummaryProvider.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Services/Room/RoomSummary/RoomSummaryProvider.swift:261]: Okay, fair 👍

File: ElementX/Resources/Localizations/en.lproj/Localizable.strings (hunk 6 above)
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Resources/Localizations/en.lproj/Localizable.strings:355]: Wait, I'm confused now, there's no `screen_invited` in localazy 🤔 

Either way, if they're in the join screen it should be prefixed with `screen_join_room`

File: ElementX/Sources/Screens/JoinRoomScreen/View/JoinRoomScreen.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Screens/JoinRoomScreen/View/JoinRoomScreen.swift:17]: Strange but okay ¯\_(ツ)_/¯

File: ElementX/Resources/Localizations/en.lproj/Localizable.strings (hunk 6 above)
[the remark this answers is not in the document]
Comment (Velin92) [reply ElementX/Resources/Localizations/en.lproj/Localizable.strings:355]: sorry it wasn't screen_invited but: `screen_invites_invited_you`
Also this is not related to the join_room but the home screen cell, is essentially the copy displayed in this new cell that will appear for knocked rooms in the room list

File: ElementX/Sources/Screens/JoinRoomScreen/View/JoinRoomScreen.swift [hunk 9]
```diff
@@ -158,6 +160,17 @@ struct JoinRoomScreen: View {
         Button(L10n.actionAccept) { context.send(viewAction: .acceptInvite) }
             .buttonStyle(.compound(.primary))
     }
+    
+    @ToolbarContentBuilder
+    private var toolbar: some ToolbarContent {
+        if context.viewState.mode == .knocked {
+            ToolbarItem(placement: .principal) {
+                RoomHeaderView(roomName: context.viewState.title,
```
[the remark this answers is not in the document]
Comment (Velin92) [reply ElementX/Sources/Screens/JoinRoomScreen/View/JoinRoomScreen.swift:168]: Well is only reused here, and only for one specific mode. Doesn't sound too bad to reuse it here. But maybe to make things easier I could make a modifier for it. The problem is that the RoomScreen one contains way more accessories, that we don't need here

File: ElementX/Sources/Services/Timeline/TimelineProxy.swift (hunk 1 above)
Comment (Velin92) [reply ElementX/Sources/Services/Timeline/TimelineProxy.swift]: done!

File: ElementX/Sources/Services/Room/RoomSummary/RoomSummaryProvider.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (Velin92) [reply ElementX/Sources/Services/Room/RoomSummary/RoomSummaryProvider.swift:261]: Actually let me improve this a bit more, I feel the `.invite` case should also have an optional `inviter` associated value

**Review by stefanceriu — APPROVED** [review]
Very nice! 👏

**Pull request merged**